DuckDuckGo/AutocompleteViewController.swift

```diff
@@ -152,47 +175,30 @@ class AutocompleteViewController: UIViewController {
     }
 
     private func cancelInFlightRequests() {
-        if let inFlightRequest = lastRequest {
-            inFlightRequest.cancel()
-            lastRequest = nil
-        }
+        task?.cancel()
+        task = nil
     }
 
     private func requestSuggestions(query: String) {
         selectedItem = -1
         tableView.reloadData()
-        do {
-            lastRequest = try AutocompleteRequest(query: query)
-            pendingRequest = true
-        } catch {
-            os_log("Couldn‘t form AutocompleteRequest for query “%s”: %s", log: .lifecycleLog, type: .debug, query, error.localizedDescription)
-            lastRequest = nil
-            pendingRequest = false
-            return
-        }
 
-        lastRequest!.execute { [weak self] (suggestions, error) in
-            guard let strongSelf = self else { return }
-
-            Task { @MainActor in
-                let matches = strongSelf.bookmarksSearch.search(query: query)
-                let notQueryMatches = matches.filter { $0.url.absoluteString != query }
-                let filteredMatches = notQueryMatches.prefix(Constants.maxLocalItems)
-                let localSuggestions = filteredMatches.map { Suggestion(source: .local,
-                                                                        suggestion: $0.title,
-                                                                        url: $0.url)
-                }
-
-                guard let suggestions = suggestions, error == nil else {
-                    os_log("%s", log: .generalLog, type: .debug, error?.localizedDescription ?? "Failed to retrieve suggestions")
-                    self?.updateSuggestions(localSuggestions)
-                    return
-                }
-
-                let combinedSuggestions = localSuggestions + suggestions
-                strongSelf.updateSuggestions(Array(combinedSuggestions))
-                strongSelf.pendingRequest = false
+        loader = SuggestionLoader(dataSource: self, urlFactory: { phrase in
+            guard let url = URL(trimmedAddressBarString: phrase),
+                  let scheme = url.scheme,
+                  scheme.description.hasPrefix("http"),
+                  url.isValid else {
+                return nil
             }
+
+            return url
+        })
+        pendingRequest = true
+
+        loader?.getSuggestions(query: query) { [weak self] result, error in
+            guard error == nil else { return }
```

Comment:
Agree on the pending request.  I think the chance of error is high (macOS seems to assume it will happen which is where I copied this from), but I'll test it and see what happens.